src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/CreditManager.cs

```diff
@@ -47,14 +49,18 @@ public ValueTask<int> RequestCreditAsync(int amount)
                     return new ValueTask<int>(granted);
                 }
 
-                var tcs = new TaskCompletionSource<int>(TaskContinuationOptions.RunContinuationsAsynchronously);
+                var tcs = new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously);
 
                 if (_waiters == null)
                 {
                     _waiters = new Queue<Waiter>();
                 }
 
-                _waiters.Enqueue(new Waiter { Amount = amount, TaskCompletionSource = tcs });
+                Waiter waiter = new Waiter { Amount = amount,
+                                             TaskCompletionSource = tcs,
+                                             TokenRegistration = cancellationToken.Register(() => tcs.TrySetCanceled(cancellationToken)) };
```

Comment:
The allocation here is slightly tricky to avoid in this case, because you need to pass in both the tcs and the cancellation token.  But that issue was already encountered and solved, via https://github.com/dotnet/corefx/blob/f7539b726c4bc2385b7f49e5751c1cff2f2c7368/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L1041.  So you'd want to make your TaskCompletionSource property actually be of that type, and then here have the registration be:
```C#
cancellationToken.Register(s => ((TaskCompletionSourceWithCancellation<int>)s).OnCanceled());
```

Reply:
Thanks. I was struggling to figure out how to do this.

Reply:
If I remember correctly, TaskCompletionSourceWithCancellation actually handles the token Register and Dispose for you. So you can just replace the line below with:

return new ValueTask<int>(tcs.WaitWithCancellationAsync(cancellationToken));